aws_lambda_powertools/utilities/idempotency/persistence/base.py

```diff
@@ -0,0 +1,441 @@
+
+        Returns
+        -------
+        bool
+            Whether the record is currently expired or not
+        """
+        if self.expiry_timestamp:
+            if int(datetime.datetime.now().timestamp()) > self.expiry_timestamp:
+                return True
+        return False
+
+    @property
+    def status(self) -> str:
+        """
+        Get status of data record
+
+        Returns
+        -------
+        str
+        """
+        if self.is_expired:
+            return STATUS_CONSTANTS["EXPIRED"]
+
+        if self._status in STATUS_CONSTANTS.values():
+            return self._status
+        else:
+            raise IdempotencyInvalidStatusError(self._status)
+
+    def response_json_as_dict(self) -> dict:
+        """
+        Get response data deserialized to python dict
+
+        Returns
+        -------
+        dict
+            previous response data deserialized
+        """
+        return json.loads(self.response_data)
+
+
+class BasePersistenceLayer(ABC):
+    """
+    Abstract Base Class for Idempotency persistence layer.
+    """
+
+    def __init__(
+        self,
+        event_key_jmespath: str = "",
+        payload_validation_jmespath: str = "",
+        expires_after_seconds: int = 60 * 60,  # 1 hour default
+        use_local_cache: bool = False,
+        local_cache_max_items: int = 256,
+        hash_function: str = "md5",
+    ) -> None:
+        """
+        Initialize the base persistence layer
+
+        Parameters
+        ----------
+        event_key_jmespath: str
+            A jmespath expression to extract the idempotency key from the event record
+        payload_validation_jmespath: str
+            A jmespath expression to extract the payload to be validated from the event record
+        expires_after_seconds: int
+            The number of seconds to wait before a record is expired
+        use_local_cache: bool, optional
+            Whether to locally cache idempotency results, by default False
+        local_cache_max_items: int, optional
+            Max number of items to store in local cache, by default 1024
+        hash_function: str, optional
+            Function to use for calculating hashes, by default md5.
+        """
+        self.event_key_jmespath = event_key_jmespath
+        if self.event_key_jmespath:
+            self.event_key_compiled_jmespath = jmespath.compile(event_key_jmespath)
+        self.expires_after_seconds = expires_after_seconds
+        self.use_local_cache = use_local_cache
+        if self.use_local_cache:
+            self._cache = LRUDict(max_items=local_cache_max_items)
+        self.payload_validation_enabled = False
+        if payload_validation_jmespath:
+            self.validation_key_jmespath = jmespath.compile(payload_validation_jmespath)
+            self.payload_validation_enabled = True
+        self.hash_function = getattr(hashlib, hash_function)
+
+    def _get_hashed_idempotency_key(self, lambda_event: Dict[str, Any]) -> str:
```

Comment:
__Question__: Should we perform this operation only once at init or is this over-optimisation?

Reply:
The persistence layer is initialized outside of the Lambda handler, so it has no awareness of the event when we create it. This method will be called with lots of different events during the instance's lifetime, so any caching would need to maintain a map of event -> key. There would be a tradeoff in terms of complexity (as dict isn't hashable so can't be used as a key by default) and memory usage. In my opinion its an unnecessary optimisation.